### PR TITLE
Fix the issue when NHWC Tensor has height or width larger then max cuda grid

### PR DESCRIPTION
**Pull request by ifedan**
When NHWC Tensor has height or width larger then max CUDA grid size, max_pool fails with error code 0

The example is: https://github.com/pytorch/pytorch/issues/28714

This change should limit grid size to the CUDA max possible size and chunk the input to be able to process it.

### CODE REVIEW

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu [hunk 1]
```diff
@@ -355,27 +357,44 @@ void max_pool2d_with_indices_out_cuda_template(
...
-              maxThreadsDim[0], std::min<int>(lastPow2(nInputPlane), at::cuda::warp_size()));
-          int block_y = std::min<int>(
-              maxThreadsDim[1], std::min<int>(lastPow2(outputWidth), max_threads / block_x));
-          int block_z = std::min<int>(
-              maxThreadsDim[2], std::min<int>(lastPow2(outputHeight), max_threads / block_x / block_y));
-          block_x = std::min<int>(
-              maxThreadsDim[0], std::min<int>(lastPow2(nInputPlane), max_threads / block_y / block_z));
-          const dim3 block(block_x, block_y, block_z);
-          int grid_x = nbatch;
-          int grid_y = cuda::ATenCeilDiv(safe_downcast<int, int64_t>(outputWidth), block_y*BLOCK_STRIDE);
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: We shouldn't need multiple kernel launching, as we are striding along width/height in the kernel already.

I would discard all changes and simply add change this line (and grid_z as well)
```
int grid_y = std::min<int>(
at::cuda::getCurrentDeviceProperties()->maxGridSize[1],
cuda::ATenCeilDiv(safe_downcast<int, int64_t>(outputWidth), block_y*BLOCK_STRIDE));
``` 

Please do *NOT* do the same thing for grid_x, as the kernel does not loop over batch dimension. The fix you have here could be used for that.

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu [hunk 2]
```diff
@@ -503,31 +522,48 @@ void max_pool2d_with_indices_backward_out_cuda_template(
...
-              maxThreadsDim[0], std::min<int>(lastPow2(nInputPlane), at::cuda::warp_size()));
-          int block_y = std::min<int>(
-              maxThreadsDim[1], std::min<int>(lastPow2(inputWidth), max_threads / block_x));
-          int block_z = std::min<int>(
-              maxThreadsDim[2], std::min<int>(lastPow2(inputHeight), max_threads / block_x / block_y));
-          block_x = std::min<int>(
-              maxThreadsDim[0], std::min<int>(lastPow2(nInputPlane), max_threads / block_y / block_z));
-          const dim3 block(block_x, block_y, block_z);
-          int grid_x = nbatch;
-          int grid_y = cuda::ATenCeilDiv(safe_downcast<int, int64_t>(inputWidth), block_y*BLOCK_STRIDE);
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: Same thing as I commented on forward.

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu (hunk 1 above)
Comment (ifedan) [reply aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: I updated grid, and removed slicing, but I still see the performance degradation for backward function.

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu (hunk 2 above)
Comment (ifedan) [reply aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: Updated

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu [hunk 3]
```diff
@@ -365,8 +365,12 @@ void max_pool2d_with_indices_out_cuda_template(
...
           const dim3 block(block_x, block_y, block_z);
           int grid_x = nbatch;
-          int grid_y = cuda::ATenCeilDiv(safe_downcast<int, int64_t>(outputWidth), block_y*BLOCK_STRIDE);
-          int grid_z = cuda::ATenCeilDiv(safe_downcast<int, int64_t>(outputHeight), block_z*BLOCK_STRIDE);
+          int grid_y = std::min<int>(
+              at::cuda::getCurrentDeviceProperties()->maxGridSize[1],
+              cuda::ATenCeilDiv(safe_downcast<int, int64_t>(outputWidth), block_y*BLOCK_STRIDE));
+          int grid_z = std::min<int>(
+              at::cuda::getCurrentDeviceProperties()->maxGridSize[2],
+              cuda::ATenCeilDiv(safe_downcast<int, int64_t>(outputHeight), block_y*BLOCK_STRIDE));
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: Typo here! `block_y` should be `block_z`

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu [hunk 4]
```diff
@@ -513,8 +517,12 @@ void max_pool2d_with_indices_backward_out_cuda_template(
...
           const dim3 block(block_x, block_y, block_z);
           int grid_x = nbatch;
-          int grid_y = cuda::ATenCeilDiv(safe_downcast<int, int64_t>(inputWidth), block_y*BLOCK_STRIDE);
-          int grid_z = cuda::ATenCeilDiv(safe_downcast<int, int64_t>(inputHeight), block_z*BLOCK_STRIDE);
+          int grid_y = std::min<int>(
+              at::cuda::getCurrentDeviceProperties()->maxGridSize[1],
+              cuda::ATenCeilDiv(safe_downcast<int, int64_t>(inputWidth), block_y*BLOCK_STRIDE));
+          int grid_z = std::min<int>(
+              at::cuda::getCurrentDeviceProperties()->maxGridSize[2],
+              cuda::ATenCeilDiv(safe_downcast<int, int64_t>(inputHeight), block_y*BLOCK_STRIDE));
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: `block_y` -> `block_z`

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu (hunk 1 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: hmmm, perf regression doesn't really make any sense here. It's basically the same kernel.
Not sure if it's related to that typo I pointed out in the other comment.

Changing the grid size would affect cache hit? I need to think about it further before I can make any decisive call. But let's fix the typo and hope the problem goes away

**Comment by ifedan** [comment]
image
image

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu (hunk 3 above)
Comment (ifedan) [reply aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: changed

File: aten/src/ATen/native/cuda/DilatedMaxPool2d.cu (hunk 4 above)
Comment (ifedan) [reply aten/src/ATen/native/cuda/DilatedMaxPool2d.cu]: changed

**Comment by jjsjann123** [comment]
LGTM.

Did that few line change actually give a hit on performance? Or is there a high variance between runs?

**Review by jjsjann123 — APPROVED** [review]

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**